### PR TITLE
Fix Caps Word not working with Auto Shift and Combo feature

### PR DESCRIPTION
**Pull request by qyurila**
## Description

Currently Caps Word feature doesn't work with Auto Shifted keys, result keys and chord keys (which are the part of Combo feature).

This PR fixes the bug by reordering event processing order, and allowing Caps Word process for user Caps Word keys' release event when Combo feature is enabled.

## Types of Changes

- [x] Bugfix
- [x] Core
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by drashna** [comment]
@getreuer just a heads up. 

Could you add additional tests for caps words, to verify that other features are working correctly?  If you're not able to do so, let me know.  But ideally, the unit tests should be expanded to verify that this is working correctly.

File: quantum/process_keycode/process_caps_word.c [hunk 1]
```diff
@@ -83,7 +83,13 @@ bool process_caps_word(uint16_t keycode, keyrecord_t* record) {
 #endif // CAPS_WORD_IDLE_TIMEOUT > 0
 
     // From here on, we only take action on press events.
-    if (!record->event.pressed) {
+    // The exceptions are the combo results and the chord keys,
```
Comment (getreuer) [inline quantum/process_keycode/process_caps_word.c:86]: At a glance, I thought that there was just kind of event generated from the combos implementation, where it calls `process_record()`, but I'm no expert. Are "combo result" and "chord key" synonymous terms, or is there a difference?

File: quantum/process_keycode/process_caps_word.c [hunk 2]
```diff
@@ -83,7 +83,13 @@ bool process_caps_word(uint16_t keycode, keyrecord_t* record) {
...
+    // which send keypress with a release event.
+    if (!record->event.pressed
+#ifdef COMBO_ENABLE
+        && !caps_word_press_user(keycode)
```
Comment (getreuer) [inline quantum/process_keycode/process_caps_word.c:90]: I'm puzzled with how this logic mismatches the comment above it: "The exceptions are the combo results and the chord keys". As it is, the code is more like "We only take action on press events OR release events where caps_word_press_user() is true," which seems a lot broader than needed.

To focus the change in behavior to combos, could something like this work?

```c
if (!record->event.pressed || IS_COMBOEVENT(record->event))
```

**Review by getreuer — CHANGES_REQUESTED** [review]

**Comment by getreuer** [comment]
@drashna Thanks for the heads up. Yes, I'm happy to add additional unit tests for this PR and also the fix for Unicode Map  https://github.com/qmk/qmk_firmware/pull/17247.

@qyurila Thanks a bunch for bringing this up and the fixes. I'd like to check with you what exactly would repro these bugs so that the new tests test the right things.

* Looking at your fix for Auto Shift, I believe the following scenario is the problem: Suppose we have Auto Shift enabled on number keys. When a number key is long-pressed, Auto Shift types a symbol. This should stop Caps Word, but before this PR it does not. Did I get that right?
* Regarding Combos, it looks like from your fix that the bugged behavior is another case of Caps Word missing the event: Suppose we have a combo that types a symbol. This should stop Caps Word, but before this PR it does not. Though, I suspect there is more to it. You mentioned in the PR description that Caps Word doesn't work with "result keys and chord keys (which are the part of Combo feature)." I don't know combos in enough detail. What is a "result key" vs. a "chord key"? If these are different things, then I ought to add a test for each. It would be helpful if you could write an outline of what some good tests would be.

File: quantum/process_keycode/process_caps_word.c (hunk 2 above)
Comment (sigprof) [reply quantum/process_keycode/process_caps_word.c:90]: Also this workaround looks really strange — the combo code should generate the same press and release events as would be generated for normal keys, and handling the release event here would result in setting weak mods on the release event too, which does not look correct. So I wonder what is the real problem here. Maybe the real bug is in the combo code, or maybe some other code calling `clear_weak_mods()` at some inappropriate time (normally `clear_weak_mods()` is called by `action_exec()` on press events before calling `action_tapping_process()` or `process_record()`, but it looks like that `dump_key_buffer()` in the combo code does that in the opposite order).

File: quantum/process_keycode/process_caps_word.c (hunk 1 above)
Comment (qyurila) [reply quantum/process_keycode/process_caps_word.c:86]: I'm so sorry, I misclicked the "Resolve conversation" button... Is there any way to undo resolving?

File: quantum/process_keycode/process_caps_word.c (hunk 2 above)
Comment (qyurila) [reply quantum/process_keycode/process_caps_word.c:90]: @getreuer I refered [this section](https://docs.qmk.fm/#/feature_combo?id=dictionary-management) of the docs for the terms "combo result" and "chord key".

For instance,

```c
//      name    result   chord keys
COMB(AB_MINS,  KC_MINS,  KC_A, KC_B)
```

when this combo is defined, `KC_MINS` will be the "combo result" and, `KC_A` and `KC_B` will be the "chord keys".
Here the problem is, not only `KC_MINS` but also `KC_A` and `KC_B` won't shifted even after turning on Caps Word.

> ```c
> if (!record->event.pressed || IS_COMBOEVENT(record->event))
> ```

In order to work, the above code should be:

```c
if (!record->event.pressed && !IS_COMBOEVENT(record->event))
```

But this code won't completely fix the problem, because `IS_COMBOEVENT()` covers only "combo result" case.
By the way, I didn't know about `IS_COMBOEVENT()` and it seems to be better to use this macro (or is it a function?) if possible.

Comment (qyurila) [reply quantum/process_keycode/process_caps_word.c:90]: @sigprof 

> the combo code should generate the same press and release events as would be generated for normal keys, and handling the release event here would result in setting weak mods on the release event too, which does not look correct.

I thought the same. But what I found on debugging the current code -- before 5ca78fd7fb2472e217b895c897abfaffd03a679e -- is that when one of the "chord keys" is tapped, a keypress is actually passed through the `caps_word_press_user()` and `true` is returned, but what is actually typed is not a shifted character. That's why I guessed that they "send keypress with a release event", and this change actually fixed the issue (strangely).

**Comment by qyurila** [comment]
@getreuer

I should have described it better. I'm sorry. What I fixed are these sinarios (sorry for duplicating my comments on conversations) :

- When we have Auto Shift enabled for letter keys, they should be shifted when Caps Word state is on, but before this PR they are not. Also, basically Mod-Tap keys are affected by Caps Word, but if Retro Shift is enabled they aren't, before this PR.

- Assume that you have this file included for your keymap:

  ```c
  // combos.def
  COMB(AB_MINS,  KC_MINS,  KC_A, KC_B)
  ```

  and not configured `caps_word_press_user()`. Then `KC_A`, `KC_B`, and `KC_MINS` -- the result of above combo -- should be shifted when Caps Word state is on, but before this PR they are not.

  - I don't mind that should the combo result keys be affected Caps Word or not, but there are [some keymaps](https://github.com/treeman/qmk_firmware/blob/master/keyboards/splitkb/kyria/keymaps/treeman/combos.def#L3-L4) which has some combos for letter keys and need it if they want to use Caps Word feature.

**Comment by getreuer** [comment]
In commit https://github.com/getreuer/qmk_firmware/commit/bfddc1e14af366eaf2dcc27c670ad52b478dd645, I wrote up a unit of using Caps Word together with Combos.

The test checks that combo chord keys and result keys work as they should for several combos, including overlapping combos and combos that chord tap-hold keys. To check thoroughly, each scenario is repeated to press the combo chord keys with a few different orders and timings.

**Surprisingly, all tests of this kind that I have tried pass** (on master and develop as of 2022-06-26). I was expecting rather that something in these tests would fail, and would be a good point to look further into this. Maybe the underlying problem has already been fixed (that would be cool), or the real problem is a scenario that hasn't been covered with this test?

**@qyurila could you please look over this test to see the combos it is checking?** Let me know if there's a particular combo definition that should be tested to trigger the bug.

* [test_caps_word_combo.cpp](https://github.com/getreuer/qmk_firmware/blob/bfddc1e14af366eaf2dcc27c670ad52b478dd645/tests/caps_word/caps_word_combo/test_caps_word_combo.cpp) – see especially the combo definitions at the top of the file.

And see also the test's configuration in these files:

* [test.mk](https://github.com/getreuer/qmk_firmware/blob/bfddc1e14af366eaf2dcc27c670ad52b478dd645/tests/caps_word/caps_word_combo/test.mk) (analogous to rules.mk)
* [config.h](https://github.com/getreuer/qmk_firmware/blob/bfddc1e14af366eaf2dcc27c670ad52b478dd645/tests/caps_word/caps_word_combo/config.h)

**Comment by qyurila** [comment]
@getreuer Thanks for your effort! It seems like the unit test covers all of the cases I encountered.

Unfortunately, the problem still exists. In conclusion, the test fails if you add this line on `test.mk`:

```c
AUTO_SHIFT_ENABLE = yes
```

So ultimately, It seems that the combo key problem occurs only when the chord keys are auto-shifted keys. (I probably mistook the condition when tweaking my keymap files. I apologize :pray: )

**Comment by getreuer** [comment]
Thanks @qyurila for pointing out to set `AUTO_SHIFT_ENABLE = yes`. With this, I am able to repro the bug.

After some printf-debugging, I found the root cause for the bug is that Combos calls `clear_weak_mods()` within `dump_key_buffer()` after passing a chord key or result key event ([code link](https://github.com/qmk/qmk_firmware/blob/a6faf3bc9e9bea48679ab17dc840adf5c029d909/quantum/process_keycode/process_combo.c#L230)), and this clears the weak Left Shift mod that Caps Word sets to capitalize the typed key.

For the above `KC_A` + `KC_B` = `KC_X` combo example, here is an outline of what happens when the chord key `KC_A` is tapped, with both Caps Word and Auto Shift turned on:

1. Combo: gets the `KC_A` press event. The event is buffered and not passed to other handlers yet.
2. Combo: gets the `KC_A` release event. The `dump_key_buffer()` function is called to pass along the tap on `KC_A`. The following all occurs within `dump_key_buffer()`...
3. `dump_key_buffer()` calls `action_tapping_process()` to pass the buffered `KC_A` press event.
4. Caps Word: gets the `KC_A` press event. The weak Left Shift mod is set to capitalize it.
5. Auto Shift: gets the `KC_A` press event. The event is buffered and further handling is skipped.
6. **⚠️ ROOT CAUSE HERE ⚠️** Execution returns from `action_tapping_process()`. `dump_key_buffer()` calls `clear_weak_mods()`, clearing the Left Shift mod that Caps Word had set.
9. `dump_key_buffer()` calls `action_tapping_process()` a second time to pass the `KC_A` release event.
10. Auto Shift: gets the `KC_A` release event and types lowercase "`a`".

A similar sequence happens with a combo result key when the `KC_A` + `KC_B` chord is tapped:

1. Combo: gets the `KC_A` press event. This event is buffered.
2. Combo: gets the `KC_B` press event. The `dump_key_buffer()` function is called to pass along the combo result key `KC_X`...
3. `dump_key_buffer()` calls `action_tapping_process()`, passing a press event on `KC_X`.
4. Caps Word: gets `KC_X` press event. The weak Left Shift mod is set to capitalize it.
5. Auto Shift: gets `KC_X` press event and buffers it.
6. **⚠️ ROOT CAUSE HERE ⚠️** Execution returns from `action_tapping_process()`. `dump_key_buffer()` calls `clear_weak_mods()`, clearing the Left Shift mod that Caps Word had set.
7. `dump_key_buffer()` calls `action_tapping_process()`, passing a release event on `KC_X`.
8. Auto Shift: gets `KC_X` release event and types lowercase "`x`".

From these outlines, it becomes clear why the bug happens with all three features Caps Word + Combos + Auto Shift used together and not with Caps Word + Combos or Caps Word + Auto Shift. 

* With Caps Word + Combos, but Auto Shift is disabled, the press event continues with default handling (and normally ultimately types a key) after Caps Word applies the weak Left Shift mod, instead of getting buffered in the Auto Shift handler. 
* With Caps Word + Auto Shift, and no Combos, the weak Left Shift mod that Caps Word sets during the press event survives until the release event, when Auto Shift types the key.

The bug can be fixed by not clearing the weak Left Shift mod. I submitted a PR https://github.com/qmk/qmk_firmware/pull/17549 to make this change.

**Comment by drashna** [comment]
Closing since other PR supersedes this
